File: pipelines/support-data/scripts/lineage_hook.py
#!/usr/bin/env python3
"""Shared helpers for automatic lineage regeneration after successful runs."""

from __future__ import annotations

import os
import subprocess
import sys
from collections.abc import Callable
from pathlib import Path
from typing import Any


ROOT_DIR = Path(__file__).resolve().parents[3]
LINEAGE_SCRIPT = ROOT_DIR / "pipelines" / "support-data" / "scripts" / "generate_data_lineage.py"
SKIP_ENV = "RETROVERSE_SKIP_LINEAGE"


def run_lineage_generation() -> None:
    if os.environ.get(SKIP_ENV):
        return
    subprocess.run([sys.executable, str(LINEAGE_SCRIPT)], check=True, cwd=ROOT_DIR)


def run_with_lineage(main_func: Callable[[], Any]) -> int:
    result = main_func()
    run_lineage_generation()
    return 0 if result is None else int(result)

File: pipelines/lib/lineage-hook.js
import { execFileSync } from 'node:child_process'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const ROOT_DIR = path.resolve(__dirname, '..', '..')
const LINEAGE_SCRIPT = path.join(ROOT_DIR, 'pipelines', 'support-data', 'scripts', 'generate_data_lineage.py')
const SKIP_ENV = 'RETROVERSE_SKIP_LINEAGE'

export function runLineageGeneration() {
  if (process.env[SKIP_ENV]) return
  execFileSync('python3', [LINEAGE_SCRIPT], {
    cwd: ROOT_DIR,
    stdio: 'inherit',
  })
}

export async function runWithLineage(task) {
  const result = await task()
  runLineageGeneration()
  return result
}
